src/3d/symbols/qgspointcloud3dsymbol_p.cpp

```diff
@@ -227,6 +267,179 @@ void QgsPointCloud3DSymbolHandler::makeEntity( Qt3DCore::QEntity *parent, const
+    {
+      filteredExtraPointData.positions.append( pos );
+      vertices.push_back( pos.x() );
+      vertices.push_back( pos.z() );
+      //vertSize += 2;
+
+      if ( hasColorData )
+        filteredExtraPointData.colors.append( outNormal.colors.at( i ) );
+      if ( hasParameterData )
+        filteredExtraPointData.parameter.append( outNormal.parameter.at( i ) );
+    }
+  }
+
+  outNormal.positions.resize( properPointsCount );
+  if ( hasColorData )
+    outNormal.colors.resize( properPointsCount );
+  if ( hasParameterData )
+    outNormal.parameter.resize( properPointsCount );
+
+  outNormal.positions.append( filteredExtraPointData.positions );
+  outNormal.colors.append( filteredExtraPointData.colors );
+  outNormal.parameter.append( filteredExtraPointData.parameter );
+
+  try
+  {
+    // Triangulation happens here
+    delaunator::Delaunator triangulation( vertices );
+
+    QVector<QVector3D> normals( outNormal.positions.count(), QVector3D( 0.0, 0.0, 0.0 ) );
+    const std::vector<size_t> &triangleIndexes = triangulation.triangles;
+    for ( size_t i = 0; i < triangleIndexes.size(); i += 3 )
+    {
+      QVector<QVector3D> triangleVertices( 3 );
+      for ( size_t j = 0; j < 3; ++j )
+      {
+        size_t vertIndex = triangleIndexes.at( i + j );
+        triangleVertices[j] = outNormal.positions.at( vertIndex );
+      }
+      //calculate normals
+      const QVector3D v1( triangleVertices.at( 1 ) - triangleVertices.at( 0 ) );
+      const  QVector3D v2( triangleVertices.at( 2 ) - triangleVertices.at( 0 ) );
+      const QVector3D partialNormal = QVector3D::crossProduct( v1, v2 );
+      for ( size_t j = 0; j < 3; ++j )
+        normals[triangleIndexes.at( i + j )] += partialNormal;
+    }
+
+    // Build now normals array
+    outNormal.normals.resize( ( outNormal.positions.count() ) * sizeof( float ) * 3 );
+    float *normPtr = reinterpret_cast<float *>( outNormal.normals.data() );
+    for ( int i = 0; i < normals.size(); ++i )
+    {
+      QVector3D normal = normals.at( i );
+      normal = normal.normalized();
+
+      *normPtr++ = normal.x();
+      *normPtr++ = normal.y();
+      *normPtr++ = normal.z();
+    }
+
+    // We apply a filter on triangles to improve the rendering:
+    // - we keep only triangles that have a least one point in the bounding box
+    // - if option is selected, we don't keep triangle with side greater than a threshold
+
+    outNormal.triangles.resize( triangleIndexes.size() * sizeof( quint32 ) );
+    quint32 *indexPtr = reinterpret_cast<quint32 *>( outNormal.triangles.data() );
+    size_t effective = 0;
+
+    bool sizeFilter = context.symbol()->filterTrianglesBySize();
+    bool heightFilter = context.symbol()->filterTrianglesByHeight();
+    float sizeThreshold =  context.symbol()->triangleSizeThreshold();
+    float heightThreshold =  context.symbol()->triangleHeightThreshold();
+
+    for ( size_t i = 0; i < triangleIndexes.size() / 3; ++i )
```

Comment:
maybe just for easier readability - previous cycle that iterates over triangleIndexes has step size of `i += 3` - that would also avoid the `i * 3` expression multiple times below, and it would be easier to read as the index `i` would have the same meaning.

`for ( size_t i = 0; i < triangleIndexes.size(); i += 3 )`